types/time.go

```diff
@@ -637,14 +638,14 @@ func parseDatetime(sc *stmtctx.StatementContext, str string, fsp int, isFloat bo
 			year = adjustYear(year)
 		case 8: // YYYYMMDD
 			_, err = fmt.Sscanf(seps[0], "%4d%2d%2d", &year, &month, &day)
```

Comment:
package main

import (
	"fmt"
)

func main() {

	var year, month, day int
	_, err := fmt.Sscanf("1991110a", "%4d%2d%2d", &year, &month, &day)

	fmt.Println(err)
}

Reply:
@zhexuany This is checked here: [ParseDateFormat](https://github.com/pingcap/tidb/blob/b30dbd06acaab088a1e8fa004e189c651ca08bf1/types/time.go#L560). If the last rune is not a number, `ParseDateFormat` would fail.

Reply:
By the way, here's another problem. In MySQL:
```sql
mysql> select cast("1991110a" as datetime);
+------------------------------+
| cast("1991110a" as datetime) |
+------------------------------+
| NULL                         |
+------------------------------+
1 row in set, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+--------------------------------------+
| Level   | Code | Message                              |
+---------+------+--------------------------------------+
| Warning | 1292 | Incorrect datetime value: '1991110a' |
+---------+------+--------------------------------------+
1 row in set (0.00 sec)
```

In TiDB:
```sql

tidb> select cast("1991110a" as datetime);
ERROR 1292 (22007): invalid time format: '1991110a'
```
That's because in [builtinCastStringAsTimeSig.evalTime](https://github.com/pingcap/tidb/blob/b30dbd06acaab088a1e8fa004e189c651ca08bf1/expression/builtin_cast.go#L1177), we don't call [handleInvalidTimeError](https://github.com/pingcap/tidb/blob/b30dbd06acaab088a1e8fa004e189c651ca08bf1/expression/errors.go#L63) to deal with invalid time format. But if we fix it here, a warning would be generated and the return value is `NULL`, thus some other builtins, such as `UNIX_TIMESTAMP`, would fail on invalid time format. For `UNIX_TIMESTAMP`, if the input is invalid, the return value is `0`,  if the input is `NULL`, it returns `NULL`, so the error must be raised, that's contradictory with the former case.

```sql
mysql> select unix_timestamp("1991110a");
+----------------------------+
| unix_timestamp("1991110a") |
+----------------------------+
|                   0.000000 |
+----------------------------+
1 row in set, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+--------------------------------------+
| Level   | Code | Message                              |
+---------+------+--------------------------------------+
| Warning | 1292 | Incorrect datetime value: '1991110a' |
+---------+------+--------------------------------------+
1 row in set (0.00 sec)

mysql> select unix_timestamp(NULL);
+----------------------+
| unix_timestamp(NULL) |
+----------------------+
|                 NULL |
+----------------------+
1 row in set (0.00 sec)

```

Maybe we should trace wether an `expression`'s children got an warning during execution. What do you think? @zhexuany @zz-jason

Reply:
@spongedu yes, it seems that mysql also uses the same way:
```cpp
1398 /**
1399    @retval true  args[0] is SQL NULL, so item is set to SQL NULL
1400    @retval false item's value is set, to 0 if out of range
1401 */
1402 bool Item_func_unix_timestamp::val_timeval(struct timeval *tm) {
1403   DBUG_ASSERT(fixed == 1);
1404   if (arg_count == 0) {
1405     tm->tv_sec = current_thd->query_start_in_secs();
1406     tm->tv_usec = 0;
1407     return false;  // no args: null_value is set in constructor and is always 0.
1408   }
1409   int warnings = 0;
1410   return (null_value = args[0]->get_timeval(tm, &warnings));
1411 }
```

and the function body for `get_timeval` is:
```cpp
1314 /**
1315    If argument is NULL, sets null_value. Otherwise:
1316    if invalid DATETIME value, or a valid DATETIME value but which is out of
1317    the supported Unix timestamp range, sets 'tm' to 0.
1318 */
1319 bool Item::get_timeval(struct timeval *tm, int *warnings) {
1320   MYSQL_TIME ltime;
1321   if (get_date(&ltime, TIME_FUZZY_DATE)) {
1322     if (null_value) return true; /* Value is NULL */
1323     goto zero;                   /* Could not extract date from the value */
1324   }
1325   if (datetime_to_timeval(current_thd, &ltime, tm, warnings))
1326     goto zero;  /* Value is out of the supported range */
1327   return false; /* Value is a good Unix timestamp */
1328 zero:
1329   tm->tv_sec = tm->tv_usec = 0;
1330   return false;
1331 }
```

Reply:
@zz-jason But it seems that the `warnings` is set, but not used here.